source/opt/ir_builder.h

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_IR_BUILDER_H_
+#define LIBSPIRV_OPT_IR_BUILDER_H_
+
+#include "opt/basic_block.h"
+#include "opt/instruction.h"
+#include "opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+// In SPIR-V, ids are encoded as uint16_t, this id is guaranteed to be always
+// invalid.
+constexpr uint32_t kInvalidId = std::numeric_limits<uint32_t>::max();
+
+// Helper class to abstract instruction construction and insertion.
+// |AnalysisToPreserve| asks the InstructionBuilder to preserve requested
+// analysis.
+// Supported analysis:
+//   - Def-use analysis
+//   - Instruction to block analysis
+template <ir::IRContext::Analysis AnalysisToPreserve =
+              ir::IRContext::kAnalysisNone>
+class InstructionBuilder {
+  static_assert(!(AnalysisToPreserve &
+                  ~(ir::IRContext::kAnalysisDefUse |
+                    ir::IRContext::kAnalysisInstrToBlockMapping)),
+                "Only Def-use analysis update is supported");
+
+ public:
+  using InsertionPointTy = spvtools::ir::BasicBlock::iterator;
+
+  InstructionBuilder(ir::IRContext* context, ir::BasicBlock* parent,
+                     InsertionPointTy insert_before)
```

Comment:
I used the iterator to allow insertion at the end of a basic block. An `Instruction*` can be turned into an InsertionPointTy, so we can still have both.

We need a Ctor that takes an `Instruction*` (insertion before this instruction) and a `BasicBlock*` (insertion at the end). That should cover the use cases and the remove the need for that extra parameter.

I will make the current Ctor internal, it can be moved as public if there is a need for this overload.